core/src/main/java/org/apache/iceberg/ManifestListWriter.java

```diff
@@ -84,23 +69,67 @@ public long length() {
     return writer.length();
   }
 
-  private static FileAppender<ManifestFile> newAppender(OutputFile file, Map<String, String> meta) {
-    try {
-      return Avro.write(file)
-          .schema(ManifestFile.schema())
-          .named("manifest_file")
-          .meta(meta)
-          .overwrite()
-          .build();
-
-    } catch (IOException e) {
-      throw new RuntimeIOException(e, "Failed to create snapshot list writer for path: " + file);
+  static class V2Writer extends ManifestListWriter {
+    private final V2Metadata.IndexedManifestFile wrapper;
+
+    V2Writer(OutputFile snapshotFile, long snapshotId, Long parentSnapshotId, long sequenceNumber) {
+      super(snapshotFile, ImmutableMap.of(
+          "snapshot-id", String.valueOf(snapshotId),
+          "parent-snapshot-id", String.valueOf(parentSnapshotId),
+          "sequence-number", String.valueOf(sequenceNumber),
+          "format-version", "2"));
+      this.wrapper = new V2Metadata.IndexedManifestFile(sequenceNumber);
+    }
+
+    @Override
+    protected ManifestFile prepare(ManifestFile manifest) {
+      return wrapper.wrap(manifest);
+    }
+
+    @Override
+    protected FileAppender<ManifestFile> newAppender(OutputFile file, Map<String, String> meta) {
+      try {
+        return Avro.write(file)
+            .schema(V2Metadata.MANIFEST_LIST_SCHEMA)
+            .named("manifest_file")
+            .meta(meta)
+            .overwrite()
+            .build();
+
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to create snapshot list writer for path: " + file);
+      }
     }
   }
 
   static class V1Writer extends ManifestListWriter {
-    private V1Writer(OutputFile snapshotFile, long snapshotId, Long parentSnapshotId) {
-      super(snapshotFile, snapshotId, parentSnapshotId);
+    private final V1Metadata.IndexedManifestFile wrapper = new V1Metadata.IndexedManifestFile();
```

Comment:
Seems like this is the place where we can pass a snapshot id similar to how we pass a sequence number to V2 and get rid of the logic for inheriting metadata for `ManifestEntry` via `setSnapshotId` and iterating through manifests in `MergingSnapshotProducer` and `FastAppend`. Do I get it correctly, @rdblue?

Reply:
Yes. I'm going to rework inheritance for snapshot ID in a separate commit.

Reply:
Great!